### PR TITLE
Add RussianRoulette drink

### PR DESCRIPTION
**Pull request by gardspirito**
## About the PR

Adds a new "Russian roulette" drink to the game. Comes with two new recipes:
1) 1u white russian + 1u black russian + 2u tricordrazine = russian roulette blank rounds.
2) 1u white russian + 1u black russian + 2u beanbags (that can be made by juicing beanbag shells) = russian roulette live rounds.
When drank, either blank or live rounds are activated, causing healing or harm respectively. Chances for each outcome are based on the amount of respective chemicals present in the solution, with ethanol contributing to the "blanks" outcome. This way, players can themselves define the risk they want to go through, and even "survive" 100% "lives" drink given they are drunk enough (have enough ethanol in their stomach) and fortune is on their side.
Activation mechanism (chemical plainly called "russian roulette") is separate from "blanks" and "lives", but is produced alongside with blanks and lives. The activation happens only when there is more than 9.5u of russian roulette in the metabolizer. Non-activated blanks and lives stay in the player and very slowly metabolize, accumulating up to 40u (40u of lives = 80 damage on activation).

Healing of blanks is ~doubled if there are at least 4u of lives present in the solution.
Lives have generic "alcohol" flavour and can be easily mixed into someone's drink, but are, again, harmless on their own.

## Why / Balance

Bartending lacks fun things. Russian roulette is a (hopefully) pro-social drink that introduces new RP situations.

## Technical details

This PR adds 4 new reagents: `RussianRouletteLive`/`RussianRouletteBlank` (slow-metabolizing drinks that accumulates up to 40u and carry no functionality by themselves), `RussianRoulette` (activator drink, which, when over 9.5u, conumes one of the former chemicals and applies respective effects once), `Beanbag` (a reagent that can be made by juicing beanbag shells and is used to create `RussianRouletteLive`).

This PR adds two new `ReagentEffects`: `ActivateRandom` and `PlaySound`.
`ActivateRandom` is a reagent effect used by the activators. For each outcome, it assesses the amount of chemicals in the solution that contribute to it, selects one of the outcome (randomly, preferring the group with greater presence in the solution), removes all non-catalyst reagents and fires the effects, providing total amount of removed chemicals as quantity to the reagent effects being run.
Each contributing chemical of each outcome can be either consumed or catalyst, and has a factor that is used to identify how much each unit contributes to the assessed presence of the entire group. For example, in case of `RussianRoulette`, `Ethanol` is considered a catalyst chemical with `0.25` factor that contributes to the "blanks" outcome.
ActivateRandom does *not* remove the activator, in case of `RussianRoulette` this is made manually by each outcome, making it one-shot reagent effect.

`PlaySound` is a reagent effect that continuously plays some sound. In case of `RussianRoulette` it is used to play a gunshot sound in the "lives" outcome.

## Media

bildo

https://github.com/space-wizards/space-station-14/assets/29687558/13a14a7e-6c01-48c1-a15d-351b33981620

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

:cl:
- add: Russian Roulette drink that can either heal or damage the drinker has been added to the game. Test your luck at the bar!

### CODE REVIEW

**Comment by gardspirito** [comment]
Code is actually ready for review, but textures for the drink are pending. They will be added later.  Marking it as "draft" for now I guess.

**Comment by gardspirito** [comment]
bildo

**Comment by UbaserB** [comment]
image

This line at the end is a bit weird

**Comment by gardspirito** [comment]
> image
> 
> This line at the end is a bit weird

I'm afraid I can't do anything about it.
It's defined in ReagentEffect base class: https://github.com/space-wizards/space-station-14/blob/e99e65929e7ac26376195a80ddb5086f30cd80ba/Content.Shared/Chemistry/Reagent/ReagentEffect.cs#L57-L69
... and is not overridable.
It just means that all effects above are activated when there is more than 9.5u (or 12.5u in the latest iteration of the PR) of RussianRoulette in the metabolizer.

**Comment by UbaserB** [comment]
I thought it was something you could edit but nvm, was mistaken. You're good.

**Comment by gardspirito** [comment]
Honestly, I have a feeling that I accidentally made one of the most complex chemicals in the game and that guidebook auto-generated explanation isn't that good at explaining of how it needs to be applied.

First of all, what could potentially be seen as a problem is that RussianRoulette is split into 3 reagents, being RussianRoulette, RussianRouletteLive and RussianRouletteBlank. It makes sense because all three can exist separately, even inside metabolizer.
1) RussianRouletteLive/RussianRouletteBlank can be present in the metabolizer without RussianRoulette. This allows for them to be accumulated in the body for a later bang.
2) RussianRoulette can be present in the metabolizer in quantities lower this 12.5u, which is a semi-intentionally allows bartender to brew the drink the wrong way. Side effects of this just include massive dehydration, nothing lethal.
... and this approach allows to investigate the contents of the drink with beer glasses.

So it makes sense (at least I think so), but can be very confusing. To brew this drink correctly, the bartender needs to mix 7.5u of White Russian with 7.5u of Black Russian with 10u of Beanbag/Tricordrazine in any proportions, but this isn't explicitly reflected anywhere.

**Comment by gardspirito** [comment]
Thank you.

> Couldn't it just be one chemical?

I believe the updated reagent system makes it possible to merge [blanks, lives] or even [blanks, lives, activator] into a single chemical. Merging just blanks and lives doesn't sound to be a problem, just no ability for players to inspect the glass with beer goggles and probably some extra work to update the guidebook entry (because it then needs to reflect that although there are two different recipes to russian roulette that produce seemingly same chemicals, it's actually not the same).
Unifying all three is possible as well, but has gameplay implications: currently, it's possible to separate chemicals via ChemMaster, if we were to unify chemicals, this isn't going to be possible any longer, so this way you can't easily accumulate lots of lives/blanks in the organism before drinking/force-feeding the activator.

> Also, having two separate recipes for the "blank" and the "live" removes randomness and kinda defeats the purpose and thrill of the gimmick.

Sure, but this again has gameplay implications:
It's no longer possible for the bartender to have any kind of control over the chances. Moreover, the current system allows adding new "ammo" into the game.

Also, I believe that instead of forcing the randomness down the throat, it's better to incentivize the player to take the risk.
In the current implementation, healing is greatly boosted if there were some Lives in the mixture.

File: Content.Server/EntityEffects/Effects/ActivateRandom.cs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+                + ":\n"
+                + string.Join("\n", case_.Effects.ToList()
+                    .Select(effect =>
+                    {
+                        var desc = effect.GuidebookEffectDescription(prototype, entSys);
+                        return desc is null ? null : "  " + desc;
+                    })
+                    .Where(x => x is not null))))
+                + "\n"));
+    }
```
Comment (metalgearsloth) [inline Content.Server/EntityEffects/Effects/ActivateRandom.cs:49]: LINQ allocates for every chain in C# and doing this by hand will probably make this a lot more readable.

File: Content.Server/EntityEffects/Effects/ActivateRandom.cs [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+                    chances[i] += impact;
+                    chancesSum += impact;
+                }
+            }
+        }
+
+        // Next, select a "random" set. Sets of higher propotions are prioritized.
+        int index = 0;
+        IRobustRandom _random = IoCManager.Resolve<IRobustRandom>();
+        {
```
Comment (metalgearsloth) [inline Content.Server/EntityEffects/Effects/ActivateRandom.cs:87]: You should avoid doing IoC resolves as they use thread-locals, especially in hot loops. Just put it as a dependency on the effect.

File: Content.Server/EntityEffects/Effects/ActivateRandom.cs [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        // some speicific set is based on how many of these chemicals (with respect to factor)
+        // are present in the solution.
+        // We use chance and chanceSum to calculate proprotions of chemical sets (with respect to factor)
+        // present in the solution.
+        float[] chances = new float[Cases.Length];
+        float chancesSum = 0f;
+
+        for (int i = 0; i < Cases.Length; i++)
+        {
+            var _case = Cases[i];
```
Comment (metalgearsloth) [inline Content.Server/EntityEffects/Effects/ActivateRandom.cs:72]: C# style guide says no prefix _.

File: Content.Server/EntityEffects/Effects/ActivateRandom.cs [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        if (args.Source is null)
+            return; // ActivateRandom activates chemicals from the args.Source.
+            // If source is not present, ActivateRandom has nothing to activate.
+
+        // Only one random set of chemicals is activated. The chance to activate
+        // some speicific set is based on how many of these chemicals (with respect to factor)
+        // are present in the solution.
+        // We use chance and chanceSum to calculate proprotions of chemical sets (with respect to factor)
+        // present in the solution.
+        float[] chances = new float[Cases.Length];
```
Comment (metalgearsloth) [inline Content.Server/EntityEffects/Effects/ActivateRandom.cs:67]: Same here we use var for styleguide.
```suggestion
        // present in the solution.
        var chances = new float[Cases.Length];
```

File: Content.Server/EntityEffects/Effects/ActivateRandom.cs [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+            }
+        }
+
+        // When some set is activated, ActivateRandom removes reagents of the set marked as "consumed".
+        // When the effects of the set are run, the amount of *consumed* reagents as passed to them.
+        // This is useful, for example, in conjuction with HealthChange's scaleByQuantity
+        FixedPoint2 consumed = 0f;
+        foreach (var activate in Cases[index].Activates) {
+            if (activate.Consume && args.Source.TryGetReagent(activate.Reagent, out var quantity)) {
+                consumed += quantity.Quantity;
```
Comment (metalgearsloth) [inline Content.Server/EntityEffects/Effects/ActivateRandom.cs:101]: as above on newline, also it should be FixedPoint2.Empty or FixedPoint2.Zero I forget which.

File: Content.Server/EntityEffects/Effects/ActivateRandom.cs [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+                _adminLogger.Add(
+                    LogType.ReagentEffect,
+                    effect.LogImpact,
+                    $"Metabolism effect {effect.GetType().Name:effect}"
+                    + $" of reagent {args.Reagent?.LocalizedName:reagent}"
+                    + $" applied on entity {args.TargetEntity:entity}"
+                    + $" at {args.EntityManager.GetComponent<TransformComponent>(args.TargetEntity).Coordinates:coordinates}"
+                );
+            }
+            effect.Effect(newArgs);
```
Comment (metalgearsloth) [inline Content.Server/EntityEffects/Effects/ActivateRandom.cs:133]: As above:
1. newline
2. Don't resolve logger inside here.
3. Styleguide for no prefix _.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/EntityEffects/Effects/ActivateRandom.cs (hunk 2 above)
Comment (gardspirito) [reply Content.Server/EntityEffects/Effects/ActivateRandom.cs]: How can this be done? I tried using 
```csharp
[Dependency] private readonly IRobustRandom _random = default!;
```
before, but it doesn't seem to work on `EntityEffect`s.

**Comment by gardspirito** [comment]
Test failure isn't related to my PR, right?..

**Comment by Winkarst-cpu** [comment]
Yeah

**Review by UbaserB — CHANGES_REQUESTED** [review]
Could you change it so that you don't need to grind beanbag shells (and hence creating an entirely new reagent) for the live rounds?

**Comment by UbaserB** [comment]
Are you coming back to this?

**Comment by gardspirito** [comment]
> Are you coming back to this?

Oh yes, sorry for the delay. In fact, I was just fixing the conflict just now.
Yes, it's easy to remove beanbag, I believe I could just drop it entirely from the recipe by allowing players to make `live rounds` via `black russian` + `white russian` mix shaking and then let them just allow to add tricordrazine into the mix to make `blank rounds`.

However, I believe there is also another problem with this PR. The problem is about metamorphic glasses: they render the name of the chemical that's present the most in the solution, and I believe it's not good when people are drinking "russian roulette live rounds glass" instead of just "russian roulette glass". I originally attempted to circumvent this by making proportions the way so that "russian roulette" chemical is always prevalent in the solution, but this is dumb, makes mixing everything harder due to strange proportions and, *takes a deep breath*, makes the drinker suffer if he doesn't drink the mix quick enough.

And I'm not sure how to fix these issues. Changing how metamorphic works so it doesn't give "russian roulette live rounds" sounds like an unnecessary change for the sake of one single chemical. I was thinking of ReagentData to make both chemicals undistinguishable, but I have a feeling it's not a suitable solution to store the quantity information.

Deleting beanbags at the moment.

**Comment by gardspirito** [comment]
Current band-aid proportions:
bildo

**Comment by Everturning** [comment]
I love this

**Comment by gardspirito** [comment]
I whipped up a patch that makes live & blank rounds non-metamorphic:
https://github.com/gardspirito/space-station-14/commit/161a333020b545214a0b903c1af2a9dbe1f9343b
bildo
This fixes the problem of ugly proportions. Although that doesn't fix the issue that slow drinking yields in bang not happening since activator metabolizes before enough is accumulated...

**Comment by gardspirito** [comment]
Lonely here.
🥲

**Comment by SlamBamActionman** [comment]
Hello! Thank you for your contribution!

We've discussed this PR internally further and while the concept of having a "russian roulette" style drink is neat, we believe that this implementation is not the way to go about it. The way to mix the drinks to get the desired effects isn't fully intuitive and it becomes its own unique system just for this drink interaction. For example a similar way to do this currently in the game using existing systems is to mix drinks in multiple opaque containers and/or use tasteless reagents (e.g. amatoxin). 

Because of this, the decision has been made to close this PR. Wish you the best in future contributions.

**Pull request closed without merging**